lib/core/src/persist/mod.rs

```diff
@@ -199,35 +198,57 @@ impl Persister {
     pub(crate) fn insert_or_update_payment_with_wallet_tx(&self, tx: &WalletTx) -> Result<()> {
         let tx_id = tx.txid.to_string();
         let is_tx_confirmed = tx.height.is_some();
-        let mut tx_balances = tx.balance.clone();
-        // Remove the Liquid Bitcoin asset balance
+        let tx_balances = tx.balance.clone();
+
+        let num_assets = tx_balances.len();
         let lbtc_asset_id = utils::lbtc_asset_id(self.network);
-        let mut balance = tx_balances
-            .remove(&lbtc_asset_id)
-            .map(|balance| (lbtc_asset_id.to_string(), balance));
-        // If the balances are still not empty pop the asset balance
-        if tx_balances.is_empty().not() {
-            balance = tx_balances
-                .pop_first()
-                .map(|(asset_id, balance)| (asset_id.to_hex(), balance));
-        }
-        let (asset_id, payment_type, amount) = match balance {
-            Some((asset_id, asset_amount)) => {
-                let payment_type = match asset_amount >= 0 {
+        let has_non_zero_foreign_asset_balances = tx_balances.iter().any(|(asset_id, b)| {
+            if *asset_id == lbtc_asset_id {
+                return false;
+            }
+            return *b != 0;
+        });
+
+        let payment_balances: Vec<PaymentTxBalance> = tx_balances
+            .into_iter()
+            .filter_map(|(asset_id, mut balance)| {
+                // https://github.com/breez/breez-sdk-liquid/pull/951#discussion_r2203640193
+                // We always compensate for fees for the outbound L-BTC balance
+                if asset_id == lbtc_asset_id
+                    && balance < 0
+                    && (num_assets == 1 || has_non_zero_foreign_asset_balances)
+                {
+                    balance += tx.fee as i64;
+
+                    // If the resulting balance is null, and we have more than one non-zero foreign asset,
+                    // we need to avoid inserting this payment balance
+                    if balance == 0 && has_non_zero_foreign_asset_balances {
```

Comment:
Why do we need the has_non_zero_foreign_asset_balances ?
In case we don't have balance for asset that is not lbtc would it make a different?